### PR TITLE
📚 Adding Knowledge llm blocks

### PR DESCRIPTION
**Pull request by abhi1092**
- 🧩 Conditional Block for conditionally picking prompt templates and parsing function
- 🎛️ Changes in the Sampling block to handle post_fixes
- 🐞 Bug fix for _parse function in LLMBlock

### CODE REVIEW

File: src/instructlab/sdg/llmblock.py [hunk 1]
```diff
@@ -21,6 +22,7 @@ def __init__(
         client,
         model_id,
         output_cols,
+        parser_kwargs={},
```
Comment (aakankshaduggal) [inline src/instructlab/sdg/llmblock.py:25]:
```suggestion
        **parser_kwargs,
```

File: src/instructlab/sdg/llmblock.py [hunk 2]
```diff
@@ -35,6 +37,9 @@ def __init__(
         self.model_prompt = model_prompt
         self.output_cols = output_cols
         self.batch_params = batch_kwargs.get("batch_kwargs", {})
+        self.parser_name = parser_kwargs.get("parser_name", None)
```
Comment (aakankshaduggal) [inline src/instructlab/sdg/llmblock.py:40]:
```suggestion
        self.parser_kwargs = parser_kwargs.get("parser_kwargs", {})
        self.parser_name = self.parser_kwargs.get("parser_name", None)
```

File: src/instructlab/sdg/llmblock.py [hunk 3]
```diff
@@ -35,6 +37,9 @@ def __init__(
...
+        self.parsing_pattern = parser_kwargs.get("parsing_pattern", None)
```
Comment (aakankshaduggal) [inline src/instructlab/sdg/llmblock.py:41]:
```suggestion
        self.parsing_pattern = self.parser_kwargs.get("parsing_pattern", None)
```

File: src/instructlab/sdg/llmblock.py [hunk 4]
```diff
@@ -35,6 +37,9 @@ def __init__(
...
+        self.parser_cleanup_tags = parser_kwargs.get("parser_cleanup_tags", None)
```
Comment (aakankshaduggal) [inline src/instructlab/sdg/llmblock.py:42]:
```suggestion
        self.parser_cleanup_tags = self.parser_kwargs.get("parser_cleanup_tags", None)
```

File: src/instructlab/sdg/llmblock.py [hunk 5]
```diff
@@ -107,3 +128,64 @@ def generate(self, samples, **gen_kwargs) -> Dataset:
...
+class ConditionalLLMBlock(LLMBlock):
+    def __init__(
+        self,
+        block_name,
+        config_paths,
+        client,
+        model_id,
+        output_cols,
+        selector_column_name,
+        parser_kwargs={},
```
Comment (aakankshaduggal) [inline src/instructlab/sdg/llmblock.py:142]:
```suggestion
        **parser_kwargs,
```

File: src/instructlab/sdg/llmblock.py [hunk 6]
```diff
@@ -107,3 +128,64 @@ def generate(self, samples, **gen_kwargs) -> Dataset:
...
+                    prompt=self.prompt_template.format(**sample).strip()
+                )
+                for sample in samples
+            ]
+        response = self.client.completions.create(
+            prompt=prompts, **{**self.defaults, **gen_kwargs}
+        )
+        return [choice.text.strip() for choice in response.choices]
+
+    def _validate(self, prompt_template: str, input_dict: Dict[str, Any]) -> bool:
```
Comment (aakankshaduggal) [inline src/instructlab/sdg/llmblock.py:188]:
```suggestion
    def validate(self, prompt_template: str, input_dict: Dict[str, Any]) -> bool:
```

**Review by aakankshaduggal — COMMENTED** [review]
Suggested some changes to pass the pylint

**Review by oindrillac — APPROVED** [review]
Tested changes in this PR, works for me!

**Review by aakankshaduggal — APPROVED** [review]
LGTM! 🚀

**Comment by russellb** [comment]
I noticed that this PR changed the format expected from the dataset given to the pipeline. The CLI integration we just merged matched the old format. (We merged it after this, I just missed that this change impacted it.) This made me realize that the skills templates expect something different, as well. I filed #55 as a follow-up to this issue.

**Pull request merged**